src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -16,32 +17,50 @@ internal class EnumConverter<T> : JsonConverter<T>
         // Odd type codes are conveniently signed types (for enum backing types).
         private static readonly string? s_negativeSign = ((int)s_enumTypeCode % 2) == 0 ? null : NumberFormatInfo.CurrentInfo.NegativeSign;
 
+        private const string ValueSeparator = ", ";
+
         private readonly EnumConverterOptions _converterOptions;
-        private readonly JsonNamingPolicy _namingPolicy;
-        private readonly ConcurrentDictionary<string, string>? _nameCache;
+        private readonly JsonNamingPolicy? _namingPolicy;
+        private readonly ConcurrentDictionary<ulong, JsonEncodedText> _nameCache;
 
         public override bool CanConvert(Type type)
         {
             return type.IsEnum;
         }
 
-        public EnumConverter(EnumConverterOptions options)
-            : this(options, namingPolicy: null)
+        public EnumConverter(EnumConverterOptions converterOptions, JsonSerializerOptions serializerOptions)
+            : this(converterOptions, namingPolicy: null, serializerOptions)
         {
         }
 
-        public EnumConverter(EnumConverterOptions options, JsonNamingPolicy? namingPolicy)
+        public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? namingPolicy, JsonSerializerOptions serializerOptions)
         {
-            _converterOptions = options;
-            if (namingPolicy != null)
-            {
-                _nameCache = new ConcurrentDictionary<string, string>();
-            }
-            else
+            _converterOptions = converterOptions;
+            _namingPolicy = namingPolicy;
+
+            string[] names = Enum.GetNames(TypeToConvert);
+            Array values = Enum.GetValues(TypeToConvert);
+            Debug.Assert(names.Length > 0 && names.Length == values.Length);
+
+            int valueCount = names.Length;
+
+            // Environment.ProcessorCount is the default concurrency level used in the concurrent dictionary implementation.
```

Comment:
Why pass it explicitly? We just want ConcurrentDictionary to be fast however it chooses to achieve that? or are we passing this because we have special knowledge about how we're going to use it and we want to pass a hint?

Reply:
These things are subtle in terms of stuff like docker containers/cgroups/NUMA etc -- one could imagine ConcurrentDictionary could be tuned in future.

Reply:
I was passing the capacity (there's currently no ctor that takes just that), but maybe that's not really great since the size would grow immediately if we see flag values post-warm up. 

Since this is warm-up/first-time (de)serialization, potentially resizing here is fine (and may even help throughput perf since we may not need to resize then). Also, ConcurrentDictionary could change as you noted. I'll just call the parameterless ctor.

PS. I'll rebase the current commit and the one I'll make to address this comment so that the major changes since the last iteration can be clear.